### PR TITLE
Algorithm for scheduling meetings between multiple people

### PR DESCRIPTION
**Pull request by AlexandraImbrisca**
Basic functionality to find the possible time slots for scheduling a meeting such as:
- All the mandatory participants will be able to attend it.
- The number of optional participants that can attend the meeting is maximised.

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/data/Marker.java [hunk 1]
```diff
@@ -14,15 +14,19 @@
 
 package com.google.sps.data;
 
-// Class used to define the basic characteristics of a marker.
+/** Class used to define the basic characteristics of a marker. */
```
Comment (nataliakrylova) [inline portfolio/src/main/java/com/google/sps/data/Marker.java:17]: Please revert changes in this file, they don't belong to this commit.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -14,10 +14,320 @@
...
+      }
+    }
+    return getTheBestAttendanceSlots(slotsAttendance);
+  }
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<String> mandatoryAttendees = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+    int duration = (int) request.getDuration();
```
Comment (nataliakrylova) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:321]: Why are you converting long to int?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -14,10 +14,320 @@
...
+   * attendees participate.
+   *
+   * @param currentEvents The events already scheduled.
+   * @param mandatoryAttendees The mandatory attendees of the desired event.
+   * @param targetDuration The duration of the desired event.
+   * @return The slots available for scheduling the meeting.
+   */
+  public List<TimeRange> getSlotsAvailable(
+      Collection<Event> currentEvents, Collection<String> mandatoryAttendees, int targetDuration) {
+    // Initialise the list of available slots
```
Comment (nataliakrylova) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:189]: Your code should be able to "speak" for itself, without additional comments. Please leave only the ones which are really necessary for understanding the code.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -14,10 +14,320 @@
...
+
+  /**
+   * Computes the number of the target attendees that also participate to the current event.
+   *
+   * @param currentEvent The event.
+   * @param targetAttendees The attendees that will be searched through the attendee list of the
+   *     currentEvent.
+   * @return The truth value of the condition described.
+   */
+  private int checkCommonAttendees(Event currentEvent, Collection<String> targetAttendees) {
```
Comment (nataliakrylova) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:89]: (nit) getNumberOfCommonAttendies

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -14,10 +14,320 @@
...
+    // Initialise the list of available slots
+    List<TimeRange> availableSlots = new ArrayList<TimeRange>();
+    // Initially, the whole day is available for scheduling the event.
+    availableSlots.add(TimeRange.WHOLE_DAY);
+
+    // For each of the events already scheduled.
+    for (Event currentEvent : currentEvents) {
+      // Check if the event is attended by at least one of the mandatory
+      // participants.
+      if (checkCommonAttendees(currentEvent, mandatoryAttendees) > 0) {
```
Comment (nataliakrylova) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:198]: In this case you can also use Collections.disjoint since you're not interested in # of common participants https://docs.oracle.com/javase/7/docs/api/java/util/Collections.html#disjoint(java.util.Collection,%20java.util.Collection)

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 6]
```diff
@@ -14,10 +14,320 @@
...
+  /**
+   * Computes the available slots after considering all the other events to which the mandatory
+   * attendees participate.
+   *
+   * @param currentEvents The events already scheduled.
+   * @param mandatoryAttendees The mandatory attendees of the desired event.
+   * @param targetDuration The duration of the desired event.
+   * @return The slots available for scheduling the meeting.
+   */
+  public List<TimeRange> getSlotsAvailable(
```
Comment (nataliakrylova) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:187]: How fast is this algorithm? Can you do better - for example, reduce the amount of removing/adding of time slots to availableSlots? Would it be possible (and if so would it be faster) to write the available time slots directly (I mean, without initially adding one for whole day and then splitting it)?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 2 above)
Comment (AlexandraImbrisca) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:321]: At the beginning I was generating intervals with that duration and I had to convert to int since the TimeRange class defines the duration as an integer. I will remove the conversion as I no longer need it. Thanks for pointing it out! :)

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 6 above)
Comment (AlexandraImbrisca) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:187]: I didn't consider the complexity of the code at this point. And good idea, I think I am going to use a hashmap to store the unavailable slots rather than keep removing them from the whole day. At the end, I will construct the available slots by considering the unavailable slots.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 7]
```diff
@@ -14,10 +14,323 @@
...
+   * @param currentEvent The event.
+   * @param targetAttendees The attendees that will be searched through the attendee list of the
+   *     currentEvent.
+   * @return The truth value of the condition described.
+   */
+  private int getNumberOfCommonAttendees(Event currentEvent, Collection<String> targetAttendees) {
+    // Create a new hashset from the event's set of attendees (since the
+    // retainAll function will remove all the attendees that are not common,
+    // we can't use a set of a fixed size).
+    Set<String> eventAttendees = new HashSet<>(currentEvent.getAttendees());
```
Comment (nataliakrylova) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:96]: Does currentEvent.getAttendees() return Collection? If so, what's the reason you need to convert it to HashSet? Afaiu retainAll exists for Collection https://docs.oracle.com/javase/8/docs/api/java/util/Collection.html#retainAll-java.util.Collection-

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 8]
```diff
@@ -14,10 +14,323 @@
...
 public final class FindMeetingQuery {
+
+  /**
+   * Initialises a list of SlotAttendance associations for a given list of time slots and an
+   * attendance of noOptionalAttendees.
+   *
+   * @param slots The given list of time slots.
+   * @param noOptionalAttendees The initial attendance of each time slot.
+   */
+  public List<SlotAttendance> initSlotsAttendance(List<TimeRange> slots, int noOptionalAttendees) {
```
Comment (nataliakrylova) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:74]: What is noOptionalAttendees? Maybe numOptionalAttendees?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 9]
```diff
@@ -14,10 +14,323 @@
...
+    // Case 2: |---------|       |---|    - mainSlot
+    //            |---|    or |---------| - minorSlot
+    //    =>   |--|   |--|    |--|   |--|
+
+    List<TimeRange> resultingSlots = new ArrayList<TimeRange>();
+
+    // Determine the common interval that should be removed.
+    TimeRange intersectionSlot = getIntersectionSlot(mainSlot, minorSlot);
+
+    int remainingLeftSlotStart = mainSlot.start();
```
Comment (nataliakrylova) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:155]: It looks like this code is based on an assumption that majorSlot always starts earlier than minor. Will it work as intended for the bottom right example if targetDuration is very small?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 10]
```diff
@@ -14,10 +14,323 @@
...
+   */
+  private List<TimeRange> getDifferenceSlots(
+      TimeRange mainSlot, TimeRange minorSlot, long targetDuration) {
+    // Based on how the slots are scheduled we can have one of the following cases:
+    //
+    // Case 1: |---|          |---|      - mainSlot
+    //            |---| or |---|         - minorSlot
+    //    =>   |--|            |--|
+    //
+    // Case 2: |---------|       |---|    - mainSlot
```
Comment (nataliakrylova) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:146]: "resulted after eliminating the minorSlot from the mainSlot" - right bottom example looks like an opposite of this.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 8 above)
Comment (AlexandraImbrisca) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:74]: I chose "no" as a short form for number. Didn't see the confusion until now :)

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 7 above)
Comment (AlexandraImbrisca) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:96]: Actually, the getAttendees() returns a Set<String>. Even though the attendees set is of type HashSet, when returning the attendees the code is written such that is transformed into an unmodifiable set.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 10 above)
Comment (AlexandraImbrisca) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:146]: My fault! I forgot to update this example. Thanks for pointing it out! The result is actually no slot since removing the common part is equal to removing the whole main slot.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 9 above)
Comment (AlexandraImbrisca) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:155]: I am not sure I understand how would this affect the code. For instance, if we use the right bottom example, the mainSlot is equal with the intersectionSlot so there are no left / right intervals remaining after eliminating the intersection slot. Maybe I am missing something, but could you please be more explicit?

Comment (nataliakrylova) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:156]: I think I got confused by bottom right example not being updated :) Now it makes more sense.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 11]
```diff
@@ -14,10 +14,324 @@
...
+
+  /**
+   * Computes the "intersection" slot - just like the intersection of two sets, the intersection
+   * slot is the common slot of two overlapsed slots.
+   *
+   * @param firstSlot The first slot.
+   * @param secondSlot The second slot.
+   * @return The intersection slot.
+   */
+  private TimeRange getIntersectionSlot(TimeRange firstSlot, TimeRange secondSlot) {
```
Comment (nataliakrylova) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:113]: What will be returned if they don't intersect?

Comment (nataliakrylova) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:113]: Please also look at what happens with getDifferenceSlots in this case.

Comment (AlexandraImbrisca) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:113]: Initially I wrote the functions, knowing that I only call them when the intervals intersect. However, I should definitely cover this case in the functions too. Thank you for pointing it out!

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 12]
```diff
@@ -270,5 +469,69 @@ public void notEnoughRoom() {
 
     Assert.assertEquals(expected, actual);
   }
-}
 
+  @Test
+  public void onlyOptionalAvailableAttendees() {
```
Comment (nataliakrylova) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:474]: The name of the text case should give an idea of what is going to happen in the test and sometimes describe the conditions shortly:)
